Improvements to specification and sampling of source

paulromano opened the pull request:
This pull request implements the changes discussed in #530, namely:

- Ability to specify multiple distributions (the relative probability for distributions is given by a "strength" attribute on a `<source>` element).
- Ability to decouple x, y, and z components of the spatial distribution.
- Ability to specify distributions for mu and phi for the angular distribution
- An abstract univariate probability distribution with a variety of extensions that can be used for any single variable (x-coordinate, mu for angle, energy)
- Tabular probability distributions (piecewise continuous as well as discrete)

The Python API has been changed as follows:

- A new openmc.stats package provides univariate and multivariate probability density classes that are used in defining the source.
- A new openmc.source module with a Source class is now used to set source distributions
- SettingsFile.source expects an iterable of type Source. As a convenience, you can set it to a scalar Source as well.

In this PR, I've also helped get rid of some module dependencies:
- string -> global
- error -> global
- random_lcg -> global

There are no test result changes. However, the fact that "strength" always is written for a source now changed all the inputs, so inputs_true.dat changed for everything.

smharper commented:
Cool PR, @paulromano!  I think this feature is going to be very powerful.  I'll give you some review this week.  You are going to add a test for it, right?

paulromano commented:
Yes, I'm going to modify the source tests to cover more of the possible cases.

paulromano commented:
Ok, I've deleted four tests and created a new one that covers all the different cases now that we can have multiple source distributions in a single run.

smharper commented:
Huh, for some reason your branch is being tested against the distribmat test even though it shouldn't exist in your branch.  Travis can be weird sometimes.  Anyway, that should go away with a merge from develop.

Regarding the code, the summary file should be modified to include the sources.

smharper commented on openmc/stats/multivariate.py:
```diff
@@ -0,0 +1,409 @@
+
+    """
+
+    def __init__(self, mu=None, phi=None, name='angle',reference_uvw=[0., 0., 1.]):
```
Space between , and "reference" (or better yet, a newline to put it under 80 columns... ya know, just in case we ever decide to punch out our Python API on Hollerith cards.)

paulromano replied:
My Hollerith cards only have 45 columns! Need to get me some of those IBM punch cards.

smharper commented on openmc/stats/multivariate.py:
```diff
@@ -0,0 +1,409 @@
+    mu : openmc.stats.Univariate
+        Distribution of the cosine of the polar angle
+    phi : openmc.stats.Univariate
+        Distribution of the azimuthal angle
```
Add "(in radians)" to this description and the one in the "Attributes" section

smharper commented on openmc/stats/multivariate.py:
```diff
@@ -0,0 +1,409 @@
+    @phi.setter
+    def phi(self, phi):
+        cv.check_type('azimuthal angle', phi, Univariate)
+        self._phi = phi
```
These `mu`s and `phi`s will only work if their `name`s are `mu` and `phi`, right?  If that's the case, we should check that condition in the setter.

smharper replied:
Even better, make the `name` attribute for `Univariate` optional, and then set them to `"mu"` and `"phi"` in these setters `if mu % name is None`

smharper commented on openmc/stats/univariate.py:
```diff
@@ -0,0 +1,329 @@
+        cv.check_type('discrete probabilities', p, Iterable, Real)
+        for pk in p:
+            cv.check_greater_than('discrete probability', pk, 0.0, True)
+        self._p = p
```
Allow `x` and `p` to be set with lone floats and ints
```python
if cv._isinstance(x, (Real, Integral)): x = [x]
```

smharper commented on openmc/stats/univariate.py:
```diff
@@ -0,0 +1,329 @@
+    def __init__(self, a, b, name='energy'):
+        super(Watt, self).__init__(name)
+        self.a = a
+        self.b = b
```
Set to 0.988 and 2.249 by default

smharper commented on src/distribution_multivariate.F90:
```diff
@@ -0,0 +1,160 @@
+!===============================================================================
+! UNITSPHEREDISTRIBUTION type defines a probability density function for points
+! on the unit sphere. Extensions of this type are used to sample angular
+! distributions for starting soures
```
typo "soures"

smharper commented on src/distribution_multivariate.F90:
```diff
@@ -0,0 +1,160 @@
+      phi = this%phi%sample()
+      uvw(:) = rotate_angle(this%reference_uvw, mu, phi)
+    end if
+  end function polar_azimuthal_sample
```
Spaces around the `%` symbols in lots of locations in this file

smharper commented on src/distribution_multivariate.F90:
```diff
@@ -0,0 +1,160 @@
+    mu = this%mu%sample()
+    if (mu == ONE) then
+      uvw(:) = this%reference_uvw
+    else
```
Do we also need a case for `if (mu == -ONE)`?

smharper commented on src/distribution_multivariate.F90:
```diff
@@ -0,0 +1,160 @@
+    integer :: i
+    real(8) :: r(3)
+
+    r = [ (prn(), i = 1,3) ]
```
Woah!  I didn't know Fortran had list comprehensions!

paulromano replied:
Yup! In Fortran parlance they're called implied DO loops.

smharper commented on src/distribution_univariate.F90:
```diff
@@ -0,0 +1,343 @@
+    if (n > 1) then
+      xi = prn()
+      c = ZERO
+      do i = 1, size(this%x)
```
Declare `i` explicitly and add `implicit none` to the new modules

smharper commented on src/distribution_univariate.F90:
```diff
@@ -0,0 +1,343 @@
+    integer :: n
+
+    ! Check interpolation parameter
+
```
Remove this blank line

smharper commented on src/distribution_univariate.F90:
```diff
@@ -0,0 +1,343 @@
+    this%c(1) = ZERO
+    do i = 2, n
+      if (this%interpolation == HISTOGRAM) then
+        this%c(i) = this%c(i-1) + this%p(i-1)*(this%x(i) - this%x(i-1))
```
If I understand correctly, the probability of landing in a certain bin `i` is `p_i * width_i`.  Wouldn't it make more sense for the probability of landing in bin `i` be simply `p_i`?

paulromano replied:
Yes, the probability of landing in a certain bin is `p_i * width_i`. The reason is that the `p_i`s then form a true probability density function in the sense that if you integrate it over the full domain you'll get one. Another way to think is that if you wanted to plot the probability density function you would do a step plot of `p_i` versus `x_i`.

smharper commented on src/distribution_univariate.F90:
```diff
@@ -0,0 +1,343 @@
+        end if
+
+      case ('watt')
+        allocate(Watt :: dist)
```
Be consistent with the space between `allocate` and the parentheses.

smharper commented on src/input_xml.F90:
```diff
@@ -339,185 +345,256 @@ subroutine read_settings_xml()
+            n = get_arraysize_double(node_angle, "parameters")
+          else
+            n = 0
+          end if
```
This `n` is not used

paulromano commented:
OK, I believe all your comments are addressed @smharper.

Regarding Travis being "weird" -- what Travis does is test a merge commit of the branch along with its target for the PR. This makes sense if you think about it; if it only tested the branch, the test results could be fine but then once its merged, tests could fail. This is exactly what happened in the case of test_distribmat. That one also needed to be updated, but it didn't exist on my branch.

@wbinventor I've updated the notebooks to use the new format. Some numbers changed, particularly OpenMOC stuff. It would be good if you could look it over and make sure the changes make sense.

wbinventor commented:
@paulromano I'll take a look at the ipython notebooks later today. I'd also like to review this if even at only a cursory level compared to @smharper, esp. the additions to the Python API.

smharper commented:
Ah, thanks for explaining the Travis behavior.  That does make sense.

How do you feel about just getting rid of the `name` attribute entirely from the Python distribution classes?  As best I can tell, setting it manually provides no benefit but gives you a failure risk.  The `Spatial` etc. object knows what names it needs so it can just directly set those XML elements.  Plus, then we could make it so that you could use the same `Univariate` e.g. for both `x` and `y` in a `SpatialIndependent`

smharper commented:
Also, consider changing `SpatialIndependent` to `CartesianIndependent` so we can later add in `CylindericalIndependent` and `SphericalIndependent`

wbinventor commented on docs/source/usersguide/input.rst:
```diff
@@ -459,67 +467,123 @@ attributes/sub-elements:
-      point source
+      point source.
+
+      For an "independent" distributoin, no parameters are specified. Instead,
```
typo: "distributoin" -> "distribution"

wbinventor commented on openmc/source.py:
```diff
@@ -0,0 +1,119 @@
+        Energy distribution of source sites
+    file : str or None
+        Source file from which sites should be sampled
+    strength : Real
```
Is it true that the `strength` is really an implied probability for a multinomial distribution over multiple `Source` objects?

wbinventor replied:
If so, must the `strengths` sum to unity or will they be normalized at runtime?

paulromano replied:
strength is an implied probability for a mass function. They are normalized at runtime so they can be whatever in the input as long as the relative strengths (probabilities) are as desired.

wbinventor commented on examples/python/reflective/build-xml.py:
```diff
@@ -1,6 +1,8 @@
 
 import openmc
+from openmc.stats import SpatialBox
+from openmc.source import Source
```
Why not add `Source` to the `__init__.py` file for `openmc`?

wbinventor commented on openmc/stats/multivariate.py:
```diff
@@ -0,0 +1,420 @@
+    phi : openmc.stats.Univariate
+        Distribution of the azimuthal angle in radians
+    name : str, optional
+        Name of the distribution. Defaults to 'angle'.
```
why not default to `'polarazimuthal'`?

wbinventor commented on openmc/stats/multivariate.py:
```diff
@@ -0,0 +1,420 @@
+    Parameters
+    ----------
+    name : str, optional
+        Name of the distribution. Defaults to 'angle'.
```
why not default to `'isotropic'`?

wbinventor commented:
@paulromano the OpenMOC results from the Jupyter Notebooks appears exactly the same to me as what is posted online. It appears that your results only differ by ~1 pcm from what is posted online. I think the likely reason is that I've installed OpenMOC compiled with double precision, whereas you probably installed with the default which is single precision.

wbinventor commented on openmc/stats/univariate.py:
```diff
@@ -0,0 +1,344 @@
+
+
+class Discrete(Univariate):
+    """Distribution characterized by a probability mass function.
```
Isn't this more formally known as a multinomial distribution?

wbinventor replied:
If so, I think a class name `Multinomial` would be more informative and transparent than `Discrete`

paulromano replied:
No, not really. A multinomial distribution gives you information on the distribution of outcomes given *n* trials. A discrete distribution is just any distribution characterized by a probability mass function (as opposed to a continuous function). We don't really care how many trials there are or what's the probability that exactly *m* out of the *n* trials results in a particular outcome (which the multinomial distribution would tell us).

wbinventor commented:
This is a very cool PR @paulromano! I've given the Python a quick run through and don't see any glaring issues beyond what I commented or queried about.

paulromano commented:
Ok, I think I've addressed all comments. @smharper I've adopted both your suggestions- changing SpatialIndependent to CartesianIndependent as well as removing the name attributes altogether. You're right that there is a lot less opportunity for user error when they don't have to worry about the name.

smharper commented:
Merged.  Thanks, @paulromano for the PR!
